pkg/kubectl/cmd/edit.go

```diff
@@ -0,0 +1,409 @@
+	}
+
+	r := resource.NewBuilder(mapper, typer, f.ClientMapperForCommand()).
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, filenames...).
+		ResourceTypeOrNameArgs(true, args...).
+		Latest().
+		Flatten().
+		Do()
+	err = r.Err()
+	if err != nil {
+		return err
+	}
+
+	infos, err := r.Infos()
+	if err != nil {
+		return err
+	}
+
+	clientConfig, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	defaultVersion := cmdutil.OutputVersion(cmd, clientConfig.Version)
+	results := editResults{}
+	for {
+		obj, err := resource.AsVersionedObject(infos, false, defaultVersion)
+		if err != nil {
+			return preservedFile(err, results.file, out)
+		}
+
+		// TODO: add an annotating YAML printer that can print inline comments on each field,
+		//   including descriptions or validation errors
+
+		// generate the file to edit
+		buf := &bytes.Buffer{}
+		if err := results.header.WriteTo(buf); err != nil {
+			return preservedFile(err, results.file, out)
+		}
+		if err := printer.PrintObj(obj, buf); err != nil {
+			return preservedFile(err, results.file, out)
+		}
+		original := buf.Bytes()
+
+		// launch the editor
+		edit := editor.NewDefaultEditor()
+		edited, file, err := edit.LaunchTempFile("kubectl-edit-", ext, buf)
+		if err != nil {
+			return preservedFile(err, results.file, out)
+		}
+
+		// cleanup any file from the previous pass
+		if len(results.file) > 0 {
+			os.Remove(results.file)
+		}
+
+		glog.V(4).Infof("User edited:\n%s", string(edited))
+		fmt.Printf("User edited:\n%s", string(edited))
+		lines, err := hasLines(bytes.NewBuffer(edited))
+		if err != nil {
+			return preservedFile(err, file, out)
+		}
+		if bytes.Equal(original, edited) {
+			if len(results.edit) > 0 {
+				preservedFile(nil, file, out)
+			} else {
+				os.Remove(file)
+			}
+			fmt.Fprintln(out, "Edit cancelled, no changes made.")
+			return nil
+		}
+		if !lines {
+			if len(results.edit) > 0 {
+				preservedFile(nil, file, out)
+			} else {
+				os.Remove(file)
+			}
+			fmt.Fprintln(out, "Edit cancelled, saved file was empty.")
+			return nil
+		}
+
+		results = editResults{
+			file: file,
+		}
+
+		// parse the edited file
+		updates, err := rmap.InfoForData(edited, "edited-file")
+		if err != nil {
+			return preservedFile(err, file, out)
+		}
+
+		visitor := resource.NewFlattenListVisitor(updates, rmap)
+
+		// need to make sure the original namespace wasn't changed while editing
+		if err = visitor.Visit(resource.RequireNamespace(cmdNamespace)); err != nil {
+			return preservedFile(err, file, out)
+		}
+
+		// use strategic merge to create a patch
+		originalJS, err := yaml.ToJSON(original)
+		if err != nil {
+			return preservedFile(err, file, out)
+		}
+		editedJS, err := yaml.ToJSON(edited)
+		if err != nil {
+			return preservedFile(err, file, out)
+		}
+		patch, err := strategicpatch.CreateStrategicMergePatch(originalJS, editedJS, obj)
+		// TODO: change all jsonmerge to strategicpatch
```

Comment:
After #13820 I want to remove all use of jsonmerge (for preconditions) and use strategicpatch pkg instead.